Tests/Realm.Tests/Sync/SessionTests.cs

```diff
@@ -221,6 +223,81 @@ public void Session_Start_IsIdempotent()
             Assert.That(session.State, Is.EqualTo(SessionState.Active));
         }
 
+        [Test]
+        public void Session_ConnectionState_FullFlow()
+        {
+            SyncTestHelpers.RunBaasTestAsync(async () =>
+            {
+                var config = await GetIntegrationConfigAsync();
+                using var realm = GetRealm(config);
+                var stateChanged = 0;
+                var completionTCS = new TaskCompletionSource<object>();
+
+                var session = realm.SyncSession;
+                session.Stop();
+
+                session.PropertyChanged += (sender, e) =>
+                {
+                    Assert.That(sender is Session, Is.True);
+                    Assert.That(e.PropertyName, Is.EqualTo(nameof(Session.ConnectionState)));
+
+                    stateChanged++;
+
+                    if (stateChanged == 1)
+                    {
+                        Assert.That(session.ConnectionState, Is.EqualTo(SessionConnectionState.Connecting));
+                    }
+                    else if (stateChanged == 2)
+                    {
+                        Assert.That(session.ConnectionState, Is.EqualTo(SessionConnectionState.Connected));
+                    }
+                    else if (stateChanged == 3)
+                    {
+                        Assert.That(session.ConnectionState, Is.EqualTo(SessionConnectionState.Disconnected));
+                        completionTCS.TrySetResult(null);
+                    }
```

Comment:
This needs a try-catch similar to what we're doing with the client reset PR otherwise an assertion will cause the test to timeout.